lib/src/stylesheet_migration.dart

```diff
@@ -0,0 +1,132 @@
+  /// path (e.g. "a" rather than "dir/a.scss").
+  final List<String> additionalUseRules = [];
+
+  /// List of patches to be applied to this file.
+  final List<Patch> patches = [];
+
+  /// Global variables in this stylesheet and its dependencies.
+  final Map<String, VariableDeclaration> variables = normalizedMap();
+
+  /// Global variables declared with !default that could be configured.
+  final Set<String> configurableVariables = normalizedSet();
+
+  /// Global mixins in this stylesheet and its dependencies.
+  final Map<String, MixinRule> mixins = normalizedMap();
+
+  /// Global functions in this stylesheet and its dependencies.
+  final Map<String, FunctionRule> functions = normalizedMap();
+
+  /// Local variables, mixins, and functions for migrations in progress.
+  ///
+  /// The migrator will modify this as it traverses the stylesheet. When at the
+  /// top-level of the stylesheet, this will be null.
+  LocalScope localScope;
+
+  StylesheetMigration._(this.stylesheet, this.path, this.contents, this.syntax);
+
+  /// Creates a new migration for the stylesheet at [path].
+  factory StylesheetMigration(String path) {
+    var contents = File(path).readAsStringSync();
+    var syntax = Syntax.forPath(path);
+    var stylesheet = Stylesheet.parse(contents, syntax, url: path);
+    return StylesheetMigration._(stylesheet, path, contents, syntax);
+  }
+
+  /// Returns the migrated contents of this file, based on [additionalUseRules]
+  /// and [patches].
+  String get migratedContents {
+    var semicolon = syntax == Syntax.sass ? "" : ";";
+    var uses = additionalUseRules.map((use) => '@use "$use"$semicolon\n');
+    var contents = Patch.applyAll(stylesheet.span.file, patches);
+    return uses.join("") + contents;
+  }
+
+  /// Declares a variable within this stylesheet, in the current local scope if
+  /// it exists, or as a global variable otherwise.
+  void declareVariable(VariableDeclaration node) {
+    if (localScope == null || node.isGlobal) {
+      variables[node.name] = node;
+      if (node.isGuarded) configurableVariables.add(node.name);
+    } else {
+      localScope.variables.add(node.name);
+    }
+  }
+
+  /// Declares a mixin within this stylesheet, in the current local scope if
+  /// it exists, or as a global mixin otherwise.
+  void declareMixin(MixinRule node) {
+    if (localScope == null) {
+      mixins[node.name] = node;
+    } else {
+      localScope.mixins.add(node.name);
+    }
+  }
+
+  /// Declares a function within this stylesheet, in the current local scope if
+  /// it exists, or as a global function otherwise.
+  void declareFunction(FunctionRule node) {
+    if (localScope == null) {
+      functions[node.name] = node;
+    } else {
+      localScope.functions.add(node.name);
+    }
+  }
+
+  /// Finds the namespace for the stylesheet containing [node], adding a new use
+  /// rule if necessary.
+  String namespaceForNode(SassNode node) {
+    var nodePath = node.span.sourceUrl.path;
```

Comment:
```suggestion
    var nodePath = p.fromUri(node.span.sourceUrl);
```

A `file:` URL's path component isn't always identical to a filesystem path.